File: FSharpWithAkka/Account/Domain/Commands.fs
namespace Bank.Account.Domain

open System
open Lib.Types

type CreateAccountCommand
   (
      entityId,
      balance: decimal,
      firstName: string,
      lastName: string,
      currency: string,
      correlationId
   ) =
   inherit Command(entityId, correlationId)
   member x.Currency = if isNull currency then "USD" else currency
   member x.Balance = balance
   member x.FirstName = firstName
   member x.LastName = lastName

type DepositCashCommand
   (entityId, amount: decimal, origin: string, correlationId) =
   inherit Command(entityId, correlationId)
   member x.Amount = amount
   member x.Origin = if isNull origin then "ATM" else origin

type DebitCommand
   (
      entityId,
      date: DateTime,
      amount: decimal,
      origin: string,
      reference: string,
      correlationId
   ) =
   inherit Command(entityId, correlationId)
   member x.Date = date
   member x.Amount = amount
   member x.Origin = origin
   member x.Reference = reference

type LimitDailyDebitsCommand(entityId, debitLimit: decimal, correlationId) =
   inherit Command(entityId, correlationId)
   member x.DebitLimit = debitLimit

type LockCardCommand(entityId, reference: string, correlationId) =
   inherit Command(entityId, correlationId)
   member x.Reference = reference

type UnlockCardCommand(entityId, reference: string, correlationId) =
   inherit Command(entityId, correlationId)
   member x.Reference = reference

File: CSharpWithLanguageExt/Account/Domain/Commands.cs
using Lib.Types;

namespace Bank.Account.Domain;

public record CreateAccountCmd(
   Guid EntityId,
   string Currency,
   decimal Balance,
   string FirstName,
   string LastName
)
: Command(EntityId)
{
   public CreatedAccount ToEvent() =>
      new(EntityId, FirstName, LastName, Timestamp, Currency, Balance);
}

public record DepositCashCmd(
   Guid EntityId,
   DateTime Date,
   decimal Amount,
   string Origin = "ATM"
)
: Command(EntityId)
{
   public DepositedCash ToEvent() => new(
      DepositedAmount: Amount,
      EntityId: EntityId,
      Timestamp: Timestamp,
      Origin: Origin
   );
}

public record DebitCmd(
   Guid EntityId,
   DateTime Date,
   decimal Amount,
   string Origin,
   string? Reference = default
)
: Command(EntityId)
{
   public DebitedAccount ToEvent() => new(
      EntityId: EntityId,
      DebitedAmount: Amount,
      Date: Date,
      Timestamp: Timestamp,
      Origin: Origin,
      Reference: Reference
   );
}

public record LimitDailyDebitsCmd(
   Guid EntityId,
   decimal DebitLimit
)
: Command(EntityId)
{
   public DailyDebitLimitUpdated ToEvent() =>
      new(EntityId, DebitLimit, Timestamp);
}

public record LockCardCmd(
   Guid EntityId,
   string Reference
)
: Command(EntityId)
{
   public LockedCard ToEvent() => new(EntityId, Timestamp, Reference);
}

public record UnlockCardCmd(
   Guid EntityId,
   string Reference
)
: Command(EntityId)
{
   public UnlockedCard ToEvent() => new(EntityId, Timestamp, Reference);
}

public record MaintenanceFeeCmd(
   Guid EntityId,
   decimal Amount = 5
)
: Command(EntityId)
{
   public MaintenanceFeeDebited ToEvent() => new(EntityId, Timestamp, Amount);
}
